### DISCUSSION TITLE
Static method for struct ?

### DISCUSSION THREAD

**Discussion by ken0x0a** [vlang/v#11935] [Ideas]
I want static method for struct

It might make json handling easier.
Currently, decoding json is depending on compiler, and `x.json2` module requires all fields to be `mut`.

(It's the same content with https://github.com/vlang/v/discussions/7610#discussioncomment-1330220)

Static method for struct is something like followings.

```vlang

enum DataType {
	str
	int_
	obj
	unknown
}
struct BaseResponse {
	typ DataType [json: @type]
	inner InnerResponse
}
struct InnerResponse {
	some string
	thing int
	important bool
}

interface Decoder<T> {
	from_json(j json2.Any) T // maybe `static from_json(j json2.Any) T` ?
}
fn decode_json<T>(src string) {
	j := json2.raw_decode(src)
	return T.from_json(j)
}
fn from_any<T>(j json2.Any) {
	return T.from_json(j)
}

fn DataType.from_json(j json2.Any) DataType {
	str := j.str()
	return match str {
		'string' { .str }
		'int', 'integer' { .int_ }
		'obj' { .obj }
		else { .unknown }
	}
}
fn BaseResponse.from_json(j json2.Any) BaseResponse {
	obj := j.as_map()
	return BaseResponse {
		inner: from_any<InnerResponse>(obj['inner'])
		typ: from_any<DataType>(obj['inner'])
	}
}
```

Syntax can be anything.

```v
fn BaseResponse.from_json(j json2.Any) BaseResponse { ... } 
```

or maybe

```vlang
fn BaseResponse::from_json(j json2.Any) BaseResponse { ... } 
```

**Comment by ken0x0a** [comment]
Now I'm creating precompile code generation script using `v.ast & v.fmt`.

And I'm using `x.json2` module but it requires all fields to be `mut`.
And also, there might be some bug (https://github.com/vlang/v/issues/11531) and I cannot implement nicely.

**Reply by ken0x0a** [reply]
https://github.com/ken0x0a/v_derive

Generated file is very ugly without static method.

**Comment by igotfr** [comment]
can be used for constructors

```v
struct Name {
  field1 int
  field2 string
}

// special syntax for constructors
fn Name.field1(field1 int) Name {
  return Name {
    field1: field1
    field2: 'default'
  }
}

fn main() {
  n := Name.field1(7)
}
```

**Comment by yashsamit** [comment]
@medvednikov This is a must feature. Any timeline estimate?

**Comment by yashsamit** [comment]
```v
//THIS is not POSSIBLE
pub struct Margin {
    top f64
    right f64
    bottom f64
    left f64
}

// if the static method allowed 
pub fn new()  Margin {}

pub type Padding = Margin
pub type Rect = Margin

Padding::new()
Margin::new()
Rect::new()


// Alternate seems ugly
// We have to add new_margin() and use the same fn for other types or create many new variants. 
// It becomes ugly when we have other constructors also like 
from_vals(a, b, c, d) 
from_val(a)

//etc


```

We can't name the module in any file because it is based on dir. I have all files inside types dir.
If the custom name of the module is allowed then it will very neat and clean.

It will like

```v
//margin and padding are module names
margin.new()
padding.new()

//Right now to achieve this, I have to create 4 folders for small files. 


```
In the doc, it mentioned that v is very modular. I feel without a filename based module name, it's not that much modular.

Solutions:
- Support custom module name based on file but not folder. 
- Add support https://github.com/vlang/v/issues/11532
- Constructor support.

**Comment by spytheman** [comment]
V now has static methods (for all types, not just structs):
```v
struct Abc { x int }
fn Abc.new() Abc { return Abc{ x: 42 } }
dump( Abc.new() )
```
produces now:
```
[a.v:3] Abc.new(): Abc{
    x: 42
}
```
